### ISSUE TITLE
Plugins with options stored on class not working on Python 3.8

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#315]
In GitLab by @adamchainz on Oct 28, 2019, 05:35

*Please describe how you installed Flake8*

Example:

```
$ pip install flake8
```

*Please provide the exact, unmodified output of `flake8 --bug-report`*

```
{
  "dependencies": [
    {
      "dependency": "entrypoints",
      "version": "0.3"
    }
  ],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.8.0",
    "system": "Darwin"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "flake8_coding",
      "version": "1.3.2"
    },
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.5.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.1.1"
    }
  ],
  "version": "3.7.8"
}
```

*Please describe the problem or feature*

Running Flake8 on macOS with flake8-coding installed ( https://github.com/tk0miya/flake8-coding/blob/master/flake8_coding.py  ) on Python 3.8 with > 1 file (so jobs are used) leads to this traceback:

```
.tox/check/bin/flake8 src/scout_apm/*.py                                                                                                                ✭
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "/Users/chainz/.pyenv/versions/3.8.0/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/chainz/Documents/Projects/scout_apm_python/.tox/check/lib/python3.8/site-packages/flake8/checker.py", line 667, in _run_checks
    return checker.run_checks()
  File "/Users/chainz/Documents/Projects/scout_apm_python/.tox/check/lib/python3.8/site-packages/flake8/checker.py", line 599, in run_checks
    self.run_ast_checks()
  File "/Users/chainz/Documents/Projects/scout_apm_python/.tox/check/lib/python3.8/site-packages/flake8/checker.py", line 503, in run_ast_checks
    for (line_number, offset, text, check) in runner:
  File "/Users/chainz/Documents/Projects/flake8-coding/flake8_coding.py", line 67, in run
    if self.encodings:
AttributeError: 'CodingChecker' object has no attribute 'encodings'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File ".tox/check/bin/flake8", line 10, in <module>
    sys.exit(main())
  File "/Users/chainz/Documents/Projects/scout_apm_python/.tox/check/lib/python3.8/site-packages/flake8/main/cli.py", line 18, in main
    app.run(argv)
  File "/Users/chainz/Documents/Projects/scout_apm_python/.tox/check/lib/python3.8/site-packages/flake8/main/application.py", line 393, in run
    self._run(argv)
  File "/Users/chainz/Documents/Projects/scout_apm_python/.tox/check/lib/python3.8/site-packages/flake8/main/application.py", line 381, in _run
    self.run_checks()
  File "/Users/chainz/Documents/Projects/scout_apm_python/.tox/check/lib/python3.8/site-packages/flake8/main/application.py", line 300, in run_checks
    self.file_checker_manager.run()
  File "/Users/chainz/Documents/Projects/scout_apm_python/.tox/check/lib/python3.8/site-packages/flake8/checker.py", line 330, in run
    self.run_parallel()
  File "/Users/chainz/Documents/Projects/scout_apm_python/.tox/check/lib/python3.8/site-packages/flake8/checker.py", line 294, in run_parallel
    for ret in pool_map:
  File "/Users/chainz/.pyenv/versions/3.8.0/lib/python3.8/multiprocessing/pool.py", line 865, in next
    raise value
AttributeError: 'CodingChecker' object has no attribute 'encodings'
```

flake8-coding uses a standard pattern to parse its options and store them as class level variables.

I think it was Python 3.8's move to use "spawn" for multiprocessing on macOS by default ( https://bugs.python.org/issue33725 ) that has caused the state to get lost between the multiprocessing jobs. Previous behaviour with "fork" mode would preserve the options parsed at the class level, but now they are lost.

I checked this theory by adding this to my `bin/flake8`:

```
import multiprocessing
multiprocessing.set_start_method("fork")
```

This fixes the traceback. But there's a reason Python moved from "fork" as the default so I'm not sure it's a good idea to keep that.

Perhaps plugins need to be fully reinitialized from (pickled?) config in each subprocess? Otherwise is there a better pattern to store plugin state?

My current workaround is `--jobs 0`.

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 28, 2019, 08:56

Hmm this actually makes sense.  We already disable `multiprocessing` on windows for the same reason (unable to pass forked state)

I'll have to think some more about how to properly do this since the current state is pretty messy (and in all likelihood need to backport at least a bugfix to disable multiprocessing while in fork mode for `flake8` 3.7.x)

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Oct 28, 2019, 09:13

I think we should make 4.0 which changes plugins and how options are passed
to them

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 28, 2019, 09:17

I agree, I'll put some thought into how this interface will look :thumbsup: -- for now I'll fix this as a bandaid

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 28, 2019, 09:22

mentioned in merge request !366

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 28, 2019, 09:50

closed via merge request !366

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 28, 2019, 09:50

closed via commit 6cc0abbea2e2f44977c7ac5e75072e0ca20c0831
